python/lsst/pipe/tasks/setPrimaryFlags.py

```diff
@@ -20,12 +20,153 @@
 #
-import numpy
+import numpy as np
 from lsst.pex.config import Config, Field, ListField
 from lsst.pipe.base import Task
 from lsst.geom import Box2D
 
 
+def getPatchInner(sources, patchInfo):
+    """Set a flag for each source if it is in the innerBBox of a patch.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+    patchInfo : `lsst.skymap.PatchInfo`
+        Information about a `SkyMap` `Patch`.
+
+    Returns
+    --------
+    isPatchInner: boolean array
+        `True` for each source that has a centroid
+        in the inner region of a patch.
+    """
+    # Extract the centroid position for all the sources
+    x = sources["slot_Centroid_x"]
+    y = sources["slot_Centroid_y"]
+    centroidFlag = sources["slot_Centroid_flag"]
+
+    # set inner flags for each source and set primary flags for sources with no children
+    # (or all sources if deblend info not available)
+    innerFloatBBox = Box2D(patchInfo.getInnerBBox())
+    inInner = innerFloatBBox.contains(x, y)
+
+    # When the centroider fails, we can still fall back to the peak, but we don't trust
+    # that quite as much - so we use a slightly smaller box for the patch comparison.
+    shrunkInnerFloatBBox = Box2D(innerFloatBBox)
+    shrunkInnerFloatBBox.grow(-1)
+    inShrunkInner = shrunkInnerFloatBBox.contains(x, y)
+
+    # Flag sources contained in the inner region of a patch
+    isPatchInner = (centroidFlag & inShrunkInner) | (~centroidFlag & inInner)
+    return isPatchInner
+
+
+def getTractInner(sources, tractInfo, skyMap):
+    """Set a flag for each source if it is contained in a given tract.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+        If `key` is not `None` then the results are written into
+        `sources[key]`.
+    tractInfo : `lsst.skymap.TractInfo`
+        Tract object
+    skyMap : `lsst.skymap.BaseSkyMap`
+        Sky tessellation object
+
+    Returns
+    -------
+    isTractInner: `bool` array
+        True for each source that has a centroid in the same
+        tract as `tractInfo`.
+    """
+    tractId = tractInfo.getId()
+    isTractInner = np.array([skyMap.findTract(s.getCoord()).getId() == tractId for s in sources])
+    return isTractInner
+
+
+def getPseudoSources(sources, pseudoFilterList, schema, log):
+    """Get a flag that marks pseudo sources.
+
+    Some categories of sources, for example sky objects,
+    are not really detected sources and should not be considered primary
+    sources.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        The catalog of sources for which to identify "pseudo"
+        (e.g. sky) objects.
+    pseudoFilterList: `list` of `str`
+        Names of filters which should never be primary
+
+    Returns
+    -------
+    isPseudo: `bool` array
```

Comment:
Spaces around ":" and I think it's supposed to be more like:
``isPseudo : array-like of `bool` ``
where "array-like" could be `` `numpy.ndarray` `` if you know that to be the case.